src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.mesos.scheduler.plan.Block;
+import org.apache.mesos.scheduler.repair.constrain.LaunchConstrainer;
+import org.apache.mesos.scheduler.repair.monitor.FailureMonitor;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides the core functionality of the automatic repair mechanisms.
+ */
+public class RepairScheduler {
+    private final Log log = LogFactory.getLog(getClass());
+
+    private final String targetConfigName;
+    private final CoreFrameworkState frameworkState;
+    private final OfferAccepter offerAccepter;
+    private final RepairOfferRequirementProvider offerReqProvider;
+    private final FailureMonitor failureMonitor;
+    private final LaunchConstrainer launchConstrainer;
+    private final AtomicReference<RepairStatus> repairStatusRef;
+
+    public RepairScheduler(
+            String targetConfigName,
+            CoreFrameworkState frameworkState,
+            RepairOfferRequirementProvider offerReqProvider,
+            OfferAccepter offerAccepter,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor,
+            AtomicReference<RepairStatus> repairStatusRef) {
+        this.targetConfigName = targetConfigName;
+        this.frameworkState = frameworkState;
+        this.offerReqProvider = offerReqProvider;
+        this.offerAccepter = offerAccepter;
+        this.failureMonitor = failureMonitor;
+        this.launchConstrainer = launchConstrainer;
+        this.repairStatusRef = repairStatusRef;
+    }
+
+    public List<OfferID> resourceOffers(SchedulerDriver driver, List<Offer> offers, Block block)
+            throws Exception {
+        List<OfferID> acceptedOffers = new ArrayList<OfferID>();
+        List<TaskInfo> terminatedTasks = getTerminatedTasks(block);
+
+        // Compute the stopped and failed pool for UI
+        RepairStatus oldStatus = repairStatusRef.get();
+        if (oldStatus == null) {
+            oldStatus = new RepairStatus(Collections.emptyList(), Collections.emptyList());
+        }
+        List<String> oldStopped = oldStatus.getStopped();
+        List<String> newStopped = new ArrayList<>(terminatedTasks.stream()
+                .filter(failureMonitor::hasFailed)
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList()));
+        newStopped.addAll(oldStopped);
+        //TODO (person) possibly write this list down in ZK, to ensure the UI doesn't change upon restart
+        newStopped = newStopped.stream().distinct().collect(Collectors.toList());
+        List<String> newFailed = terminatedTasks.stream()
+                .filter(it -> !failureMonitor.hasFailed(it))
+                .map(TaskInfo::getName)
+                .collect(Collectors.toList());
+
+        OfferRequirement offerReq = null;
+        boolean replacingMissing = false;
+
+        if (terminatedTasks.size() > 0) {
```

Comment:
I'm sorry, this is out of scope at this point. We've been collaborating on this design for two weeks. I showed you this API on July 21st in [this commit](https://github.com/mesosphere/dcos-kafka-service/pull/218/commits/6827267f4872d3601f74ca628ec9b9c04bd4a8e5), and you didn't comment on it then.

Changing this would require redesigning the control flow of the RepairScheduler and the API for the RepairOfferRequirementProvider, which will change the delivery timeline.